x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/action/GetDataFrameJobsAction.java

```diff
@@ -0,0 +1,208 @@
+import org.elasticsearch.action.Action;
+import org.elasticsearch.action.ActionRequestBuilder;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.FailedNodeException;
+import org.elasticsearch.action.TaskOperationFailure;
+import org.elasticsearch.action.support.tasks.BaseTasksRequest;
+import org.elasticsearch.action.support.tasks.BaseTasksResponse;
+import org.elasticsearch.client.ElasticsearchClient;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.xpack.ml.featureindexbuilder.job.FeatureIndexBuilderJob;
+import org.elasticsearch.xpack.ml.featureindexbuilder.job.FeatureIndexBuilderJobConfig;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+public class GetDataFrameJobsAction extends Action<GetDataFrameJobsAction.Response>{
+
+    public static final GetDataFrameJobsAction INSTANCE = new GetDataFrameJobsAction();
+    public static final String NAME = "cluster:monitor/xpack/feature_index_builder/get";
+    public static final ParseField JOBS = new ParseField("jobs");
+    public static final ParseField COUNT = new ParseField("count");
+
+    private GetDataFrameJobsAction() {
+        super(NAME);
+    }
+
+    @Override
+    public Response newResponse() {
+        return new Response();
+    }
+
+    public static class Request extends BaseTasksRequest<Request> implements ToXContent {
+        private String id;
+
+        public Request(String id) {
+            if (Strings.isNullOrEmpty(id) || id.equals("*")) {
+                this.id = MetaData.ALL;
+            } else {
+                this.id = id;
+            }
+        }
+
+        public Request() {}
+
+        @Override
+        public boolean match(Task task) {
+            // If we are retrieving all the jobs, the task description does not contain the id
+            if (id.equals(MetaData.ALL)) {
+                return task.getDescription().startsWith(FeatureIndexBuilderJob.PERSISTENT_TASK_DESCRIPTION_PREFIX);
+            }
+            // Otherwise find the task by ID
+            return task.getDescription().equals(FeatureIndexBuilderJob.PERSISTENT_TASK_DESCRIPTION_PREFIX + id);
+        }
+
+        public String getId() {
+            return id;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            super.readFrom(in);
+            id = in.readString();
+            if (Strings.isNullOrEmpty(id) || id.equals("*")) {
```

Comment:
This shouldn't be necessary since the object that was serialized should already have had this logic applied.  (Also, using `writeString` in the `writeTo` method means it's impossible for the string that's read to be `null`.)